OemPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -0,0 +1,136 @@
+/** @file
+  This application will attempt to append the 'Windows UEFI CA 2023' and then reboot the system.
+  On success, this application will allow the system to boot into 2023 signed Windows
+
+  Copyright (C) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/UefiApplicationEntryPoint.h>
+#include <Guid/ImageAuthentication.h>
+
+#include "RecoveryPayload.h"
+
+// 10 seconds in microseconds
+#define STALL_10_SECONDS  10000000
+
+/**
+  The user Entry Point for Application. The user code starts with this function
+  as the real entry point for the application.
+
+  @param[in] ImageHandle        The firmware allocated handle for the EFI image.
+  @param[in] SystemTable        A pointer to the EFI System Table.
+
+  @retval EFI_SUCCESS           The entry point is executed successfully.
+  @retval EFI_INVALID_PARAMETER SystemTable provided was not valid.
+  @retval other                 Some error occurs when executing this entry point.
+
+**/
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      Attributes;
+
+  Attributes = EFI_VARIABLE_NON_VOLATILE |
+               EFI_VARIABLE_BOOTSERVICE_ACCESS |
+               EFI_VARIABLE_RUNTIME_ACCESS |
+               EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+               EFI_VARIABLE_APPEND_WRITE;
+
+  //
+  // Start checking that the system is in a state we can safely use
+  //
+  if (SystemTable == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->ConOut == NULL) || (SystemTable->ConOut->OutputString == NULL) || (SystemTable->ConOut->ClearScreen == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->BootServices == NULL) || (SystemTable->BootServices->Stall == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // After this point, we should be able to print and stall but nothing else has been verified
+  //
+  if ((SystemTable->RuntimeServices == NULL) || (SystemTable->RuntimeServices->ResetSystem == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  //
+  // Start informing the user of what is happening
+  //
+  Status = SystemTable->ConOut->ClearScreen (SystemTable->ConOut);
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  Status = SystemTable->ConOut->OutputString (SystemTable->ConOut, L"\r\nAttempting to update the system's secureboot certificates\r\n");
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }

```

Comment:
For this and the above clear screen (and I guess for checking if these functions exist), do we actually want to fail out? Sure, the system would be in a weird state, but if all we can't do is print, seemingly it would be better to fix the certificate so the system could boot again?

Reply:
I think an `ASSERT()` would probably be fine. Developers will know something is wrong if they're testing and hit but in scenarios asserts are taken out anyway, it won't apply.

Reply:
Including an ASSERT would require including more dependencies. I don't think that it buys enough to justify it. I think I'm going to remove the status checks and just allow them to fail since they would be removed in release anyways.  If there are no complaints this is what I plan to do

Reply:
I think that's okay for these ConOut cases.